cli/src/errors/error.rs

```diff
@@ -1,49 +1,92 @@
-#[cfg(backtrace)]
-use std::backtrace::Backtrace;
 use std::convert::From;
+use std::error::Error;
+use std::fmt;
 
 use anyhow::Result;
 use log::error;
-use thiserror::Error;
 
+use crate::errors::ErrorDetails;
 use crate::errors::ExitCode;
 
-#[derive(Error, Debug)]
-#[error("{msg}")]
+#[derive(Debug)]
 pub struct ApolloError {
-    msg: String,
-    #[cfg(backtrace)]
-    backtrace: Backtrace,
-    /// The result of `error.exit_code()`.
-    exit_code: ExitCode,
+    inner: Box<Inner>,
+}
+
+#[derive(Debug)]
+struct Inner {
+    kind: ErrorDetails,
+    source: Option<Box<dyn Error>>,
 }
 
 #[allow(dead_code)]
 impl ApolloError {
     /// Returns the process exit code that should be returned if the process exits with this error.
     pub fn exit_code(&self) -> ExitCode {
-        self.exit_code
+        self.inner.kind.exit_code()
+    }
+
+    pub fn from_source<E>(source: E, kind: ErrorDetails) -> Self
+    where
+        E: Into<Box<dyn Error>>,
+    {
+        ApolloError {
+            inner: Box::new(Inner {
+                kind,
+                source: Some(source.into()),
+            }),
+        }
+    }
+
+    pub fn kind(&self) -> &ErrorDetails {
+        &self.inner.kind
     }
 }
 
-/// The failure trait for all Apollo errors.
-pub trait ApolloFail: std::error::Error {
-    /// Returns the process exit code that should be returned if the process exits with this error.
-    fn exit_code(&self) -> ExitCode;
+impl fmt::Display for ApolloError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.inner.kind.fmt(f)
+    }
 }
 
-impl<T: ApolloFail> From<T> for ApolloError {
-    fn from(error: T) -> Self {
-        let exit_code = error.exit_code();
+impl Error for ApolloError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        self.inner.source.as_ref().map(|b| b.as_ref())
+    }
+}
+
+impl From<ErrorDetails> for ApolloError {
+    fn from(kind: ErrorDetails) -> Self {
         ApolloError {
-            msg: error.to_string(),
-            exit_code,
+            inner: Box::new(Inner { kind, source: None }),
         }
     }
 }
 
+pub trait Context<T> {
+    fn with_context<F>(self, f: F) -> Fallible<T>
```

Comment:
I don't see anything calling this, would prefer to avoid dead code, unless there's a concrete use case coming up? Is this used in a later PR?

(I suspect there is no warning because the trait is public.)

Reply:
Happy to remove this. I mainly took it from https://docs.rs/anyhow/1.0.31/anyhow/trait.Context.html#tymethod.with_context. I've used `with_context` a few times but have always ended up refactoring to not need it or use `map_err`. I *think* this is better since its lazy but again, I'm mainly trying to replicate what others have done 😅

Reply:
`with_context` means you don't have to handle the source of the error, instead we can cast it for you.

So `some_result_fn().with_context(|| ErrorDetails::CustomError)?` over `some_result_fn().map_error(|e| ErrorDetails::CustomError)?` 

I *think* the latter looses the source information of the error

Reply:
_Generally_ I prefer adding something the first time it's used, or if I have concrete plans to use it soon. While it might seem like a good idea, it's easy to add it when we need it.